{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ac0cc08",
   "metadata": {},
   "source": [
    "# RNA-DNAm: scBOND"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4043325e",
   "metadata": {},
   "source": [
    "The following tutorial demonstrate how to use scBOND model making translation between scRNA-seq and scDNAm profiles."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf9d3187",
   "metadata": {},
   "source": [
    "There are three parts of this tutorial:\n",
    "\n",
    "**(1) Read in data and load it to scBOND.** This part will tell you how to load and pre-process scRNA-seq and scDNAm profiles data for scBOND model.\n",
    "\n",
    "**(2) Construct a scBOND model.** This part will tell you how to generate and train a scBOND model.\n",
    "\n",
    "**(3) Get prediction and evaluate the performance.** This part will tell you how to get prediction from scBOND model and evaluate the performance of prediction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80aba232",
   "metadata": {},
   "source": [
    "## Read in data and load it to scBOND"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6709e774",
   "metadata": {},
   "source": [
    "scBOND has integrated the process of pre-processing, constructing model, training model and evaluating in scBOND.bond for paired scRNA-seq and scDNAm profiles. You could easily use it as follow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df08ee0a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:03:39.811953Z",
     "start_time": "2025-08-20T03:03:36.675628Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/home/methy2rna/anaconda3/envs/scB/lib/python3.9/site-packages/bamnostic/__init__.py:41: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      " UserWarning:/home/methy2rna/anaconda3/envs/scB/lib/python3.9/site-packages/torch/nn/_reduction.py:42: size_average and reduce args will be deprecated, please use reduction='mean' instead.\n"
     ]
    }
   ],
   "source": [
    "import scBond\n",
    "from scBond.bond import Bond\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "805937e8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:03:39.822474Z",
     "start_time": "2025-08-20T03:03:39.815520Z"
    }
   },
   "outputs": [],
   "source": [
    "bond = Bond()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f722d24",
   "metadata": {},
   "source": [
    "Here we use the [**HumanBrainA data**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140493) with 4013 nuclei isolated from postmortem human frontal cortex(Luo, C., et al., 2023) for train and the [**HumanBrainB data**](https://datasets.cellxgene.cziscience.com/b8bedb33-4f07-41b4-9656-8c428d549c94.h5ad) compriseing 3142 cells from the human cerebral cortex for test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3786e72",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:03:52.809903Z",
     "start_time": "2025-08-20T03:03:41.697116Z"
    }
   },
   "outputs": [],
   "source": [
    "RNA_data = sc.read_h5ad('adata_rna.h5ad')\n",
    "MET_data = sc.read_h5ad('adata_met.h5ad')\n",
    "RNA_test = sc.read_h5ad('test_rna.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9939acb4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:04:58.134717Z",
     "start_time": "2025-08-20T03:04:58.122709Z"
    }
   },
   "outputs": [],
   "source": [
    "#ensure that RNA and DNAm data are paired\n",
    "rna_index_order = RNA_data.obs.index\n",
    "MET_data = MET_data[rna_index_order, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "879b2ff0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:05:03.473494Z",
     "start_time": "2025-08-20T03:05:03.395912Z"
    }
   },
   "outputs": [],
   "source": [
    "#make the indices of the two RNA datasets var the same, facilitating concatenation\n",
    "new_index = [idx.split('.')[0] for idx in RNA_data.var.index]\n",
    "RNA_data.var.index = new_index\n",
    "RNA_data = RNA_data[:, ~RNA_data.var.index.duplicated(keep='first')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee04e65f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:05:44.946009Z",
     "start_time": "2025-08-20T03:05:37.572738Z"
    }
   },
   "outputs": [],
   "source": [
    "RNA_data.obs['cell_type'] = RNA_data.obs['MajorType']\n",
    "RNA_data = RNA_data.concatenate(RNA_test, join='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "93898b2d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:05:58.047840Z",
     "start_time": "2025-08-20T03:05:50.523262Z"
    }
   },
   "outputs": [],
   "source": [
    "all_indices = MET_data.var.index.tolist()\n",
    "MET_data.var['chrom'] = [idx[:4] if len(idx) >= 4 else idx for idx in all_indices]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a392a115",
   "metadata": {},
   "source": [
    "We provide data split settings in scBOND.split_dataset for reproducibility. You could use parameters train_id, validation_id and test_id customize your own datasets splitting settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5dd44da1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:06:01.189168Z",
     "start_time": "2025-08-20T03:06:01.164298Z"
    }
   },
   "outputs": [],
   "source": [
    "train_id = [i for i in range(len(MET_data.obs_names))]\n",
    "test_id = [i+len(MET_data.obs_names) for i in range(len(RNA_test.obs_names))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eddeda47",
   "metadata": {},
   "source": [
    "Then we load data into scBOND."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dbf0952c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:06:07.012140Z",
     "start_time": "2025-08-20T03:06:03.379104Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] Bond: successfully load in data with\n",
      "\n",
      "RNA data:\n",
      "AnnData object with n_obs × n_vars = 35078 × 54928\n",
      "    obs: 'fig2_tsne_0', 'fig2_tsne_1', 'fig2_umap_0', 'fig2_umap_1', 'MajorType', 'SubType', 'cell_type', 'roi', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'supercluster_term', 'cluster_id', 'subcluster_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_id', 'is_primary_data', 'tissue_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'batch'\n",
      "    var: 'Biotype-1', 'Chromosome-1', 'End-1', 'Gene-1', 'Start-1', 'feature_is_filtered-1', 'feature_name-1', 'feature_reference-1', 'feature_biotype-1', 'feature_length-1', 'feature_type-1'\n",
      "\n",
      "MET data:\n",
      "AnnData object with n_obs × n_vars = 4013 × 309556\n",
      "    obs: 'fig2_tsne_0', 'fig2_tsne_1', 'fig2_umap_0', 'fig2_umap_1', 'MajorType', 'SubType'\n",
      "    var: 'chrom'\n"
     ]
    }
   ],
   "source": [
    "bond.load_data(RNA_data, MET_data, train_id, test_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad584488",
   "metadata": {},
   "source": [
    "scBOND could pre-process data Automatically, custom settings is also provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8b9fd4e0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:11:05.192573Z",
     "start_time": "2025-08-20T03:06:07.013896Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] RNA preprocessing: normalize size factor.\n",
      "[INFO] RNA preprocessing: log transform RNA data.\n",
      "[INFO] RNA preprocessing: choose top 3000 genes for following training.\n",
      "[INFO] MET preprocessing: Filtering features: min 29.0 cells (0.700%) per feature\n",
      "[INFO] MET preprocessing: Performing missing value imputation using method: median\n",
      "[INFO] MET preprocessing: Performing min-max scaling.\n"
     ]
    }
   ],
   "source": [
    "bond.data_preprocessing()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3927d54",
   "metadata": {},
   "source": [
    "## Construct a scBOND model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61e94e56",
   "metadata": {},
   "source": [
    "We use no data augmentation strategy here, for more information, see in scBOND-Aug."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "492dd37e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:11:11.741025Z",
     "start_time": "2025-08-20T03:11:11.736406Z"
    }
   },
   "outputs": [],
   "source": [
    "bond.augmentation(enable_augmentation=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b4ecfa6",
   "metadata": {},
   "source": [
    "scBOND model needs the information of peaks chromosomes, we should sort the scDNAm profiles data with chromosomes and tell scBOND the count of peaks in each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c8464fcd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:11:17.308289Z",
     "start_time": "2025-08-20T03:11:13.080079Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[22471, 23784, 19482, 18767, 17739, 16740, 15537, 14281, 11873, 94526, 12268, 14783, 2012, 1]"
     ]
    }
   ],
   "source": [
    "chrom_list = []\n",
    "last_one = ''\n",
    "for i in range(len(bond.MET_data_p.var.chrom)):\n",
    "    temp = bond.MET_data_p.var.chrom[i]\n",
    "    if temp[0 : 3] == 'chr':\n",
    "        if not temp == last_one:\n",
    "            chrom_list.append(1)\n",
    "            last_one = temp\n",
    "        else:\n",
    "            chrom_list[-1] += 1\n",
    "    else:\n",
    "        chrom_list[-1] += 1\n",
    "        \n",
    "print(chrom_list, end=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1c061985",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:11:17.317222Z",
     "start_time": "2025-08-20T03:11:17.310484Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "284264"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(chrom_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3767512",
   "metadata": {},
   "source": [
    "We construct and train scBOND model with default settings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b8ca00db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T03:11:20.273792Z",
     "start_time": "2025-08-20T03:11:17.318651Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] Bond: successfully construct bond model.\n"
     ]
    }
   ],
   "source": [
    "bond.construct_model(chrom_list=chrom_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "466e880e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T04:00:53.362433Z",
     "start_time": "2025-08-20T03:11:20.276378Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] Bond: training bond model ...\n",
      "[INFO] Trainer: MET pretraining ...\n",
      "MET pretrain: 100%|█████████████████████| 100/100 [09:49<00:00,  5.90s/it, train=0.2708, val=0.2739]\n",
      "[INFO] Trainer: RNA pretraining ...\n",
      "RNA pretrain: 100%|█████████████████████| 100/100 [08:36<00:00,  5.17s/it, train=0.1372, val=0.1445]\n",
      "[INFO] Trainer: Integrative training ...\n",
      "Integrative training: 100%|█████████████| 200/200 [31:05<00:00,  9.33s/it, train=1.3556, val=1.3499]\n"
     ]
    }
   ],
   "source": [
    "bond.train_model()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41d9e010",
   "metadata": {},
   "source": [
    "## Get prediction and evaluate the performance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92458143",
   "metadata": {},
   "source": [
    "You could get cross-modal predictions using bond.test_model with default settings. We also provided more information metrics in this function, see in API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9fe25efd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:01.389333Z",
     "start_time": "2025-08-20T04:03:11.006057Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] SingleModalPredictor: predicting methylation from RNA data...\n",
      "RNA to Methylation predicting...: 100%|███████████████████████████| 486/486 [00:20<00:00, 23.81it/s]\n"
     ]
    }
   ],
   "source": [
    "R2M_predict = bond.predict_single_modal(data_type='rna', load_model = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:scB]",
   "language": "python",
   "name": "conda-env-scB-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
